File: fastapiApp/routers/temperature.py
from fastapi import APIRouter, Request, Form
from fastapi.responses import HTMLResponse
from fastapi.templating import Jinja2Templates
from services.temperature import convert_temperature

templates = Jinja2Templates(directory="templates")

router = APIRouter()
@router.get("/", response_class=HTMLResponse)
async def temp_form(request: Request):
    return templates.TemplateResponse(
        "index.html",
        {"request": request, "result": None, "units": ["Celsius", "Fahrenheit", "Kelvin"], "active_tab": "temperature"}
    )

@router.post("/", response_class=HTMLResponse)
async def temp_convert(
    request: Request,
    value: float = Form(...),
    from_unit: str = Form(...),
    to_unit: str = Form(...),
):
    result = f"{value} {from_unit} = {round(convert_temperature(value, from_unit, to_unit), 2)} {to_unit}"
    return templates.TemplateResponse(
        "index.html",
        {"request": request, "result": result, "units": ["Celsius", "Fahrenheit", "Kelvin"], "active_tab": "temperature"}
    )

File: golang/services/temperature.go
package services

import "fmt"

func ConvertTemperature(value float64, fromUnit, toUnit string) (float64, error) {
	var celsius float64

	switch fromUnit {
	case "Celsius":
		celsius = value
	case "Fahrenheit":
		celsius = (value - 32) * 5 / 9
	case "Kelvin":
		celsius = value - 273.15
	default:
		return 0, fmt.Errorf("unsupported fromUnit: %s", fromUnit)
	}

	switch toUnit {
	case "Celsius":
		return celsius, nil
	case "Fahrenheit":
		return (celsius*9/5 + 32), nil
	case "Kelvin":
		return celsius + 273.15, nil
	default:
		return 0, fmt.Errorf("unsupported toUnit: %s", toUnit)
	}
}
